File: user_study_2/data/lineplot/lineplot.py
import json
import matplotlib.pyplot as plt
from matplotlib.ticker import MaxNLocator
import os
import pandas as pd


def get_filepath(name):
    current_file_path = os.path.abspath(__file__)
    directory = os.path.dirname(current_file_path)
    return os.path.join(directory, name)


def plot(dataframe):
    # Plot the 'iOS' column
    fig = plt.figure(figsize=(10, 6))
    plt.plot(dataframe['Date'], dataframe['Apple'], marker='.')
    plt.suptitle("Apple's Market Share Worldwide", x="0.26", fontsize="12")
    plt.title("From 2010 to 2023", loc="left", fontsize="10")
    plt.xlabel("Year-Month")
    plt.ylabel('Market Share (%)')
    plt.grid(True)
    txt = "Source: Mobile vendor market share from https://gs.statcounter.com/vendor-market-share/mobile"
    plt.figtext(0.5, 0.001, txt, horizontalalignment='left', fontsize=8)
    # plt.xticks(rotation=45)  # Rotate the x-axis labels for better readability

    # Set the x-axis major locator to reduce the number of ticks
    ax = plt.gca()  # Get the current axis
    ax.xaxis.set_major_locator(MaxNLocator(nbins=6))  # Set maximum number of x-axis bins

    # Save the plot to SVG
    output = get_filepath("apple_market_share_alltime_lineplot.svg")
    plt.savefig(output, format='svg')
    plt.show()  # Display the plot


def maidr(dataframe):
    # Construct the JSON structure
    maidr_data = {
        "type": "line",
        "id": "line",
        "title": "Apple's Market Share Worldwide",
        "subtitle": "From 2010 to 2023",
        "caption": "Source: Mobile vendor market share from https://gs.statcounter.com/vendor-market-share/mobile",
        "selector": "g[id='line2d_25'] > path",
        "axes": {
            "x": {
                "label": "Year-Month"
            },
            "y": {
                "label": "Market Share (%)"
            }
        },
        "data": [
            {"x": year, "y": share} for year, share in zip(dataframe['Date'], dataframe['Apple'])
        ]
    }

    # Convert the dictionary to JSON
    output = get_filepath("lineplot_alltime_maidr_data.json")
    with open(output, 'w') as json_file:
        json.dump(maidr_data, json_file, indent=4)


if __name__ == '__main__':
    # Load the dataset
    filename = get_filepath("mobile-vendor-share-alltime.csv")
    df = pd.read_csv(filename)
    plot(df)
    maidr(df)

File: user_study_2/data/lineplot/lineplot.r
# Load the necessary libraries
library(ggplot2)
library(gridSVG)
library(here)
library(jsonlite)

# Read the dataset from the CSV file
path <- here("..", "dataset", "mobile-os-share-us.csv")
market_share_data <- read.csv(path)

market_share_data$Date <- as.Date(as.character(market_share_data$Date), format = "%Y")

# We can use ggplot2 to create the line plot.
ios_market_share_plot <- ggplot(market_share_data, aes(x = Date, y = iOS)) +
  geom_line() + # This adds the line to the plot
  geom_point() + # This adds points to each data point on the line
  scale_x_date(date_labels = "%Y", date_breaks = "1 year") + # This formats the x-axis labels as years and sets the breaks
  labs(title = "iOS Market Share Over Time", x = "Year", y = "Market Share (%)") +
  theme_minimal() # This sets a minimal theme for the plot

# Draw the plot in an off-screen graphics device
print(ios_market_share_plot)

# Export the plot to SVG
grid.export("ios_market_share.svg")

# Extract the year from the Date column (assuming it's in 'YYYY' format as a string)
market_share_data$Date <- format(as.Date(market_share_data$Date, format="%Y"), "%Y")
# Construct the JSON structure
json_data <- list(
  type = "line",
  id = "line",
  title = "",
  axes = list(
    x = list(label = ""),
    y = list(label = "")
  ),
  data = lapply(seq_len(nrow(market_share_data)), function(i) {
    list(x = market_share_data$Date[i], y = market_share_data$`iOS`[i])
  })
)

# Convert the list to JSON
json_plot_data <- toJSON(json_data, pretty = TRUE, auto_unbox = TRUE)

# Print the JSON data
cat(json_plot_data)